File: problems/28-fenwick-tree/001-range-sum-query-fenwick/java/Solution.java
public class Solution {
    private final int[] bit;
    private final int[] values;

    public Solution(int[] nums) {
        values = nums.clone();
        bit = new int[nums.length + 1];
        for (int i = 0; i < nums.length; i++) add(i + 1, nums[i]);
    }

    private void add(int index, int delta) {
        while (index < bit.length) {
            bit[index] += delta;
            index += index & -index;
        }
    }

    private int prefixSum(int index) {
        int sum = 0;
        while (index > 0) {
            sum += bit[index];
            index -= index & -index;
        }
        return sum;
    }

    public void update(int index, int value) {
        int delta = value - values[index];
        values[index] = value;
        add(index + 1, delta);
    }

    public int sumRange(int left, int right) {
        return prefixSum(right + 1) - prefixSum(left);
    }
}

File: problems/28-fenwick-tree/001-range-sum-query-fenwick/python/solution.py
class NumArray:
    def __init__(self, nums: list[int]):
        self.values = nums[:]
        self.bit = [0] * (len(nums) + 1)
        for i, value in enumerate(nums, 1):
            self._add(i, value)

    def _add(self, index: int, delta: int) -> None:
        while index < len(self.bit):
            self.bit[index] += delta
            index += index & -index

    def _prefix_sum(self, index: int) -> int:
        total = 0
        while index > 0:
            total += self.bit[index]
            index -= index & -index
        return total

    def update(self, index: int, val: int) -> None:
        delta = val - self.values[index]
        self.values[index] = val
        self._add(index + 1, delta)

    def sumRange(self, left: int, right: int) -> int:
        return self._prefix_sum(right + 1) - self._prefix_sum(left)
